tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/APIInvokeWithOAuth2andBasicAuthTestCase.java

```diff
@@ -0,0 +1,185 @@
+
+    @BeforeClass
+    public void start() throws Exception {
+        String label = "apimTestLabel";
+        String project = "apimTestProject";
+        //get mock APIM Instance
+        MockAPIPublisher pub = MockAPIPublisher.getInstance();
+        API api = new API();
+        api.setName("PizzaShackAPI");
+        api.setContext("/pizzashack");
+        api.setProdEndpoint(getMockServiceURLHttp("/echo/prod"));
+        api.setSandEndpoint(getMockServiceURLHttp("/echo/sand"));
+        api.setVersion("1.0.0");
+        api.setProvider("admin");
+        //Register API with label
+        pub.addApi(label, api);
+
+        //Define application info
+        ApplicationDTO application = new ApplicationDTO();
+        application.setName("jwtApp");
+        application.setTier("Unlimited");
+        application.setId((int) (Math.random() * 1000));
+
+        //Register a production token with key validation info
+        KeyValidationInfo info = new KeyValidationInfo();
+        info.setApi(api);
+        info.setApplication(application);
+        info.setAuthorized(true);
+        info.setKeyType(TestConstant.KEY_TYPE_PRODUCTION);
+        info.setSubscriptionTier("Unlimited");
+        //set security schemas
+        String security = "oauth2,basic";
+        //Register a production token with key validation info
+        prodToken = pub.getAndRegisterAccessToken(info);
+
+        //Register a sandbox token with key validation info
+        KeyValidationInfo infoSand = new KeyValidationInfo();
+        infoSand.setApi(api);
+        infoSand.setApplication(application);
+        infoSand.setAuthorized(true);
+        infoSand.setKeyType(TestConstant.KEY_TYPE_SANDBOX);
+        infoSand.setSubscriptionTier("Unlimited");
+        sandToken = pub.getAndRegisterAccessToken(infoSand);
+
+        jwtTokenProd = getJWT(api, application, "Unlimited", TestConstant.KEY_TYPE_PRODUCTION, 3600);
+        jwtTokenSand = getJWT(api, application, "Unlimited", TestConstant.KEY_TYPE_SANDBOX, 3600);
+        expiringJwtTokenProd = getJWT(api, application, "Unlimited", TestConstant.KEY_TYPE_PRODUCTION, 1);
+        //generate apis with CLI and start the micro gateway server
+        super.init(label, project,security);
+    }
+
+    @Test(description = "Test API invocation with a oauth token")
+    public void testApiInvoke() throws Exception {
+        //test prod endpoint
+        invoke(prodToken, MockHttpServer.PROD_ENDPOINT_RESPONSE, 200);
+
+        //test sand endpoint
+        invoke(sandToken, MockHttpServer.SAND_ENDPOINT_RESPONSE, 200);
+    }
+
+    @Test(description = "Test API invocation with a JWT token")
+    public void testApiInvokeWithJWT() throws Exception {
+        //test prod endpoint with jwt token
+        invoke(jwtTokenProd, MockHttpServer.PROD_ENDPOINT_RESPONSE, 200);
+
+        //test sand endpoint
+        invoke(jwtTokenSand, MockHttpServer.SAND_ENDPOINT_RESPONSE, 200);
+
+        try {
+            Thread.sleep(2000);
```

Comment:
Why do we need a Thread.sleep here?

Reply:
It's to get JWT token  expired and assert that in next invoke.